syntax = "proto3";

package proto;

// Session Service Protocol Buffer definitions

service SessionService {
  // Get existing session or create a new one
  rpc GetOrCreateSession(GetOrCreateSessionRequest) returns (GetOrCreateSessionResponse);
  
  // Add messages to a session
  rpc AddMessages(AddMessagesRequest) returns (AddMessagesResponse);
  
  // Get messages from a session
  rpc GetMessages(GetMessagesRequest) returns (GetMessagesResponse);
  
  // Delete a session
  rpc DeleteSession(DeleteSessionRequest) returns (DeleteSessionResponse);
  
  // Save memory entry
  rpc SaveMemory(SaveMemoryRequest) returns (SaveMemoryResponse);
  
  // Get memory entries
  rpc GetMemory(GetMemoryRequest) returns (GetMemoryResponse);
  
  // Delete a memory entry
  rpc DeleteMemory(DeleteMemoryRequest) returns (DeleteMemoryResponse);
  
  // Clear all memories for a user
  rpc ClearUserMemory(ClearUserMemoryRequest) returns (ClearUserMemoryResponse);
}

// Session management messages
message GetOrCreateSessionRequest {
  string user_id = 1;
  optional string session_id = 2;  // Optional: pass to resume existing, omit for new
}

message GetOrCreateSessionResponse {
  string session_id = 1;
  string user_id = 2;
  int64 created_at = 3;
  int64 updated_at = 4;
}

// Message operations
message AddMessagesRequest {
  string session_id = 1;
  repeated Message messages = 2;
}

message AddMessagesResponse {
  bool success = 1;
  int32 message_count = 2;
}

message GetMessagesRequest {
  string session_id = 1;
  optional int32 limit = 2;
  optional int32 offset = 3;
  optional string strategy = 4;  // Context window strategy: "truncate", "hierarchical", etc.
}

message GetMessagesResponse {
  repeated Message messages = 1;
  int32 total_count = 2;
}

message DeleteSessionRequest {
  string session_id = 1;
}

message DeleteSessionResponse {
  bool success = 1;
}

// Message structure
message Message {
  string role = 1;  // "user", "assistant", "system", or "tool"
  optional string content = 2;
  repeated MessageToolCall tool_calls = 3;
  optional string tool_call_id = 4;
  optional string name = 5;
  int64 timestamp = 6;
}

message MessageFunction {
  string name = 1;
  string arguments = 2;  // JSON-encoded string
}

message MessageToolCall {
  string id = 1;
  string type = 2;  // "function"
  MessageFunction function = 3;
}

// Memory operations
message SaveMemoryRequest {
  string user_id = 1;
  string key = 2;
  string value = 3;  // JSON-encoded value
  optional string session_id = 4;
}

message SaveMemoryResponse {
  bool success = 1;
}

message GetMemoryRequest {
  string user_id = 1;
  optional string key = 2;  // If provided, get specific key; otherwise get all
  optional string session_id = 3;  // Filter by session if provided
}

message GetMemoryResponse {
  map<string, string> memories = 1;  // key -> JSON-encoded value
}

message DeleteMemoryRequest {
  string user_id = 1;
  string key = 2;
  optional string session_id = 3;
}

message DeleteMemoryResponse {
  bool success = 1;
}

message ClearUserMemoryRequest {
  string user_id = 1;
}

message ClearUserMemoryResponse {
  int32 count = 1;  // Number of memories deleted
}

// Backwards compatibility - for existing SDK code
message Session {
  string session_id = 1;
  string user_id = 2;
  int64 created_at = 3;
  int64 updated_at = 4;
}
